services/navigation/builtin/builtin_test.go

```diff
@@ -285,201 +498,421 @@ func TestStartWaypoint(t *testing.T) {
+		defer teardown()
+
+		// add waypoint 1
+		pt1 := geo.NewPoint(1, 0)
+		err := ns.AddWaypoint(ctx, pt1, nil)
+		test.That(t, err, test.ShouldBeNil)
 
-		pt1, pt2, pt3 := geo.NewPoint(1, 2), geo.NewPoint(2, 3), geo.NewPoint(3, 4)
-		points := []*geo.Point{pt1, pt2, pt3}
-		t.Run("MoveOnGlobe error results in skipping the current waypoint", func(t *testing.T) {
-			// Set manual mode to ensure waypoint loop from prior test exits
-			err = ns.SetMode(ctx, navigation.ModeManual, map[string]interface{}{"experimental": true})
-			test.That(t, err, test.ShouldBeNil)
-			ctx, cancelFunc := context.WithCancel(ctx)
-			defer ns.(*builtIn).activeBackgroundWorkers.Wait()
-			defer cancelFunc()
-			err = deleteAllWaypoints(ctx, ns)
-			for _, pt := range points {
-				err = ns.AddWaypoint(ctx, pt, nil)
-				test.That(t, err, test.ShouldBeNil)
-			}
+		// test that added waypoints are returned by the waypoints function
+		wps, err := ns.Waypoints(ctx, nil)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(wps), test.ShouldEqual, 1)
+		wpSimilar(t, wps[0], pointToWP(pt1))
+
+		// set mode to Waypoint to begin MoveOnMap calls
+		err = ns.SetMode(ctx, navigation.ModeWaypoint, nil)
+		test.That(t, err, test.ShouldBeNil)
 
-			ns.(*builtIn).startWaypoint(ctx, map[string]interface{}{"experimental": true})
+		expectedMotionCfg := &motion.MotionConfiguration{
+			LinearMPerSec:     1,
+			AngularDegsPerSec: 1,
+		}
 
-			// Get the ID of the first waypoint
-			wp1, err := ns.(*builtIn).store.NextWaypoint(ctx)
-			test.That(t, err, test.ShouldBeNil)
+		timeoutCtx, cancelFn := context.WithTimeout(context.Background(), time.Second)
+		defer cancelFn()
+
+		// confirm MoveOnGlobe is called for waypoint 1
+		expectedCall1and2 := moveOnGlobeReq{
+			componentName:      b.Name(),
+			destination:        pt1,
+			heading:            math.NaN(),
+			movementSensorName: injectMSensor.Name(),
+			motionCfg:          expectedMotionCfg,
+			// TODO: fix with RSDK-4583
+			// extra defaults to "motion_profile": "position_only"
+			// extra: map[string]interface{}{"motion_profile": "position_only"},
+		}
+		// MoveOnGlobe called twice with same parameters, as nav services
+		// retries first call which returned an error
+		blockTillCalledOrTimeout(timeoutCtx, t, expectedCall1and2, callChan)
+		blockTillCalledOrTimeout(timeoutCtx, t, expectedCall1and2, callChan)
+
+		// confirm waypoint 1 is eventually removed from the waypoints store after
+		// MoveOnGlobe reports reaching waypoint 1
+		wps1, err := pollTillWaypointLen(timeoutCtx, t, ns, 0)
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, len(wps1), test.ShouldEqual, 0)
+	})
 
-			// Reach the first waypoint
-			eventChannel <- arrivedAtWaypointMsg
-			test.That(t, <-statusChannel, test.ShouldEqual, arrivedAtWaypointMsg)
-			currentInputsShouldEqual(ctx, t, kinematicBase, pt1)
+	t.Run("MoveOnGlobe success false results in retrying the same waypoint until success", func(t *testing.T) {
```

Comment:
optionally can collapse this and the one above into a single one that checks if the `moveResponse` is `Failed`.  Would have to reintroduce the function you had written initially

Reply:
discussed offline, not possible based on current status